### PR TITLE
gh-107265: Fix initialize/remove_tools for ENTER_EXECUTOR case

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-107265

### CODE REVIEW

**Comment by corona10** [comment]
@gvanrossum 
See why this fix is needed: https://github.com/python/cpython/issues/107265#issuecomment-1693405607
But I am not sure how I can write test code for this case :(

File: Python/instrumentation.c [hunk 1]
```diff
@@ -711,7 +717,21 @@ remove_tools(PyCodeObject * code, int offset, int event, int tools)
...
     assert(event != PY_MONITORING_EVENT_INSTRUCTION);
     assert(PY_MONITORING_IS_INSTRUMENTED_EVENT(event));
-    assert(opcode_has_event(_Py_GetBaseOpcode(code, offset)));
+    _Py_CODEUNIT co_instr = _PyCode_CODE(code)[offset];
+    uint8_t co_code = co_instr.op.code;
+    uint8_t co_arg = co_instr.op.arg;
+    if (co_code == ENTER_EXECUTOR) {
+        _PyExecutorObject *exec = code->co_executors->executors[co_arg];
+        assert(exec != NULL);
+        assert(exec->vm_data.opcode != ENTER_EXECUTOR);
```
Comment (gvanrossum) [inline Python/instrumentation.c]: Why these two asserts? In the previous chunk you seem to trust the executor.

File: Python/instrumentation.c [hunk 2]
```diff
@@ -575,6 +580,7 @@ de_instrument(PyCodeObject *code, int i, int event)
         opcode_ptr = &code->_co_monitoring->per_instruction_opcodes[i];
         opcode = *opcode_ptr;
     }
+    assert(opcode != ENTER_EXECUTOR);
```
Comment (gvanrossum) [inline Python/instrumentation.c]: Maybe move this back to just below the `if` block you added? That's effectively where it was before.

File: Python/instrumentation.c [hunk 3]
```diff
@@ -711,7 +717,21 @@ remove_tools(PyCodeObject * code, int offset, int event, int tools)
...
+        assert(exec != NULL);
+        assert(exec->vm_data.opcode != ENTER_EXECUTOR);
+        co_code = _PyOpcode_Deopt[exec->vm_data.opcode];
+        co_arg = exec->vm_data.oparg;
+    }
+    else {
+        co_code = _Py_GetBaseOpcode(code, offset);
+    }
+    assert(co_code != ENTER_EXECUTOR);
+    assert(opcode_has_event(co_code));
```
Comment (gvanrossum) [inline Python/instrumentation.c]: It's unfortunate that all that code just goes to compute `co_opcode` which is only used in these two asserts. Maybe enclose in `#ifndef NDEBUG` / `#endif`?

**Review by gvanrossum — COMMENTED** [review]
I don't know this code very well; let's see if @markshannon wants to review it.

File: Python/instrumentation.c [hunk 4]
```diff
@@ -711,7 +717,22 @@ remove_tools(PyCodeObject * code, int offset, int event, int tools)
...
+            assert(exec->vm_data.opcode != ENTER_EXECUTOR);
+            co_code = _PyOpcode_Deopt[exec->vm_data.opcode];
+            co_arg = exec->vm_data.oparg;
+        }
+        else {
+            co_code = _Py_GetBaseOpcode(code, offset);
+        }
+        assert(co_code != ENTER_EXECUTOR);
+        assert(opcode_has_event(co_code));
+    #endif
```
Comment (gvanrossum) [inline Python/instrumentation.c]: Left indent the whole block please. (Our use of `#if[def]` in the rest of the code base is different from how we do it in bytecodes.c.)

File: Python/instrumentation.c [hunk 5]
```diff
@@ -711,7 +717,22 @@ remove_tools(PyCodeObject * code, int offset, int event, int tools)
     assert(event != PY_MONITORING_EVENT_LINE);
     assert(event != PY_MONITORING_EVENT_INSTRUCTION);
     assert(PY_MONITORING_IS_INSTRUMENTED_EVENT(event));
-    assert(opcode_has_event(_Py_GetBaseOpcode(code, offset)));
+    #ifndef NDEBUG
+        _Py_CODEUNIT co_instr = _PyCode_CODE(code)[offset];
+        uint8_t co_code = co_instr.op.code;
+        uint8_t co_arg = co_instr.op.arg;
```
Comment (gvanrossum) [inline Python/instrumentation.c]: Could we rename these two variables to `opcode` and `oparg`? The `co_` prefix makes me think these are fields of code objects.

**Review by gvanrossum — APPROVED** [review]
LG except formatting/naming nits.

**Review by gvanrossum — APPROVED** [review]
Green light! Thanks so much for powering through these.

File: Python/instrumentation.c [hunk 6]
```diff
@@ -566,7 +566,13 @@ de_instrument(PyCodeObject *code, int i, int event)
     _Py_CODEUNIT *instr = &_PyCode_CODE(code)[i];
     uint8_t *opcode_ptr = &instr->op.code;
     int opcode = *opcode_ptr;
-    assert(opcode != ENTER_EXECUTOR);
```
Comment (markshannon) [inline Python/instrumentation.c:569]: This assertion is correct. Please don't remove assertions, unless you are really sure that they are incorrect.

`ENTER_EXECUTOR` should never have associated instrumentation.

File: Python/instrumentation.c [hunk 7]
```diff
@@ -711,7 +717,22 @@ remove_tools(PyCodeObject * code, int offset, int event, int tools)
     assert(event != PY_MONITORING_EVENT_LINE);
     assert(event != PY_MONITORING_EVENT_INSTRUCTION);
     assert(PY_MONITORING_IS_INSTRUMENTED_EVENT(event));
-    assert(opcode_has_event(_Py_GetBaseOpcode(code, offset)));
```
Comment (markshannon) [inline Python/instrumentation.c:714]: This is also correct, provided the assertion `assert(PY_MONITORING_IS_INSTRUMENTED_EVENT(event))` is true.

File: Python/instrumentation.c [hunk 8]
```diff
@@ -1282,9 +1303,16 @@ initialize_tools(PyCodeObject *code)
...
+        int oparg = instr->op.arg;
+        if (opcode == ENTER_EXECUTOR) {
+            _PyExecutorObject *exec = code->co_executors->executors[oparg];
+            opcode = exec->vm_data.opcode;
+            oparg = exec->vm_data.oparg;
+        }
+        else if (opcode == INSTRUMENTED_LINE) {
             opcode = code->_co_monitoring->lines[i].original_opcode;
         }
+        assert(opcode != ENTER_EXECUTOR);
```
Comment (markshannon) [inline Python/instrumentation.c:1315]: This assert should be moved before the original `if (opcode == INSTRUMENTED_LINE) {`, we shouldn't get here with and `ENTER_EXECUTOR` present.

**Review by markshannon — COMMENTED** [review]
All the removed `opcode != ENTER_EXECUTOR` assertions were correct.
There should never be instrumented `ENTER_EXECUTOR` instructions.

I think the correct fix is to remove all `ENTER_EXECUTOR` instructions in `_Py_Instrument()` prior to calling `update_instrumentation_data()`, some thing like:
```
if (code->co_executors->size > 0) {
    // Walk code removing ENTER_EXECUTOR
    // Clear all executors
}
```

**Comment by markshannon** [comment]
Note that it is OK to have instrumented instructions and `ENTER_EXECUTOR` in the same code objects, but only if `ENTER_EXECUTOR` is inserted *after* the instrumentation happened, and `ENTER_EXECUTOR` never replaces an instrumented instruction.

File: Python/instrumentation.c (hunk 7 above)
Comment (corona10) [reply Python/instrumentation.c:714]: Hmm, it will guarantee that the opcode is not ENTER_EXECUTOR?

File: Python/instrumentation.c (hunk 8 above)
Comment (corona10) [reply Python/instrumentation.c:1315]: Move it to L1310 will be enough?

File: Python/instrumentation.c (hunk 6 above)
Comment (corona10) [reply Python/instrumentation.c:569]: It was moved to L574, it will be the same effect no?

**Comment by corona10** [comment]
> I think the correct fix is to remove all ENTER_EXECUTOR instructions in _Py_Instrument() prior to calling update_instrumentation_data(), some thing like:

I will try to apply this approach :)

**Pull request merged**